metrics.py

```diff
@@ -443,72 +459,176 @@ def __init__(self, exogenous_attr: Collection):
+            # the features are linearly dependent.
+            eigen_vals, eigen_vecs = np.linalg.eigh(features_cov_mat)
+            small_eigen_vals_indices = np.where(np.abs(eigen_vals) < thr)[0]
+            if small_eigen_vals_indices.size == 0:
+                is_linearly_dependent = False
+            else:
+                # For each linear dependency (corresponding to an eigen vector with a low value), mark the indices of
+                # features that are involved (identified by a non-zero coefficient in the eigen vector).
+                dependent_features_flags = np.zeros((small_eigen_vals_indices.size, self.num_of_features))
+                for i in range(small_eigen_vals_indices.size):
+                    dependent_features_flags[
+                        i, np.where(np.abs(eigen_vecs[:, small_eigen_vals_indices[i]]) > thr)[0]] = 1
+
+                # Calculate the fraction of dependencies each feature is involved in.
+                fraction_of_dependencies_involved = dependent_features_flags.mean(axis=0)
+
+                # Sort the features (their indices) by the fraction of dependencies they are involved in (remove first
+                # features that are involved in more dependencies). Break ties by the original order of the features in
+                # the array (for the consistency of sorting. E.g, if we need to get rid of degree features, always
+                # remove them by the order of nodes).
+                removal_order = np.lexsort((np.arange(self.num_of_features), -fraction_of_dependencies_involved))
+
+                # Iterate the metrics to find one with multiple features, namely effective number of features that is
+                # larger than 1 ('trimmable'). We prefer to trim metrics rather than totally eliminate them from the
+                # collection.
+                i = 0
+                cur_metric = self.get_metric_by_feat_idx(removal_order[i])
+                is_trimmable = cur_metric._get_effective_feature_count() > 1
+                while (not is_trimmable and i < removal_order.size - 1 and fraction_of_dependencies_involved[
+                    removal_order[i]] > 0):
+                    i += 1
+                    cur_metric = self.get_metric_by_feat_idx(removal_order[i])
+                    is_trimmable = cur_metric._get_effective_feature_count() > 1
+
+                # If a trimmable metric was not found (i.e., all features that are involved in the dependency are of
+                # metrics with an effective number of features of 1), totally remove the metric that is involved in most
+                # dependencies.
+                if not is_trimmable:
+                    first_metric = self.get_metric_by_feat_idx(removal_order[0])
+                    print(f"Removing the metric {str(first_metric)} from the collection to fix multi-collinearity")
+                    self.metrics = tuple([m for m in self.metrics if str(m) != str(first_metric)])
```

Comment:
I wrote a simple test case for metrics = [NumEdges(), NumEdges()] to check that removing metrics that aren't trimmable.

What ended up happening was that both of them were removed, instead of just one. I guess that's it's happening because of this line

Reply:
I guess you're right, but isn't that an invalid case? I mean - a user that gives identical metrics as inputs deserves the worst error we get think of.
Maybe we can just add a validation for that before, I don't think that the collinearity fixer should deal with this case.

Reply:
but this will also happen if we have `node_attribute_sum_out` twice, no?

I agree that NumEdges that appears twice is a degenerate scenario. My concern is that removing a metric by it's name can cause trouble, if we have the same metric twice for different use-cases (which can happen with exogenous data).